docs/overview.rst

```diff
@@ -0,0 +1,350 @@
+  general approach is to have an API which is "as simple as possible" while
+  not violating the other goals.
+
+
+Universal binaries
+  It is possible to compile extensions to a single binary which is
+  ABI-compatible across multiple Python versions and/or multiple
+  implementation. See :ref:`hpy-target-abis`.
+
+
+Opt-in low level data structures
+  Internal details might still be available, but in a opt-in way: for example,
+  if Cython wants to iterate over a list of integers, it can ask if the
+  implementation provides a direct low-level access to the content (e.g. in
+  the form of a ``int64_t[]`` array) and use that. But at the same time, be
+  ready to handle the generic fallback case.
+
+
+API vs ABI
+-----------
+
+HPy defines *both* and API and an ABI. Before digging further into details,
+let's distinguish them:
+
+  - The **API** works at the level of source code: it is the set of functions,
+    macros, types and structs which developers can use to write their own
+    extension modules.  For C programs, the API is generally made available
+    through one or more header files (``*.h``).
+
+  - The **ABI** works at the level of compiled code: it is the interface between
+    the host interpreter and the compiled DLL.  Given a target CPU and
+    operating system it defines things like the set of exported symbols, the
+    precise memory layout of objects, the size of types, etc.
+
+In general it is possible to compile the same source into multiple compiled
+libraries, each one targeting a different ABI. :pep:`3149` states that the
+filename of the compiled extension should contain the *ABI tag* to specificy
+what is the target ABI. For example, if you compile an extension called
+``simple.c`` on CPython 3.7, you get a DLL called
+``simple.cpython-37m-x86_64-linux-gnu.so``:
+
+  - ``cpython-37m`` is the ABI tag, in this case CPython 3.7
+
+  - ``x86_64`` is the CPU architecture
+
+  - ``linux-gnu`` is the operating system
+
+The same source code compiled on PyPy3.6 7.2.0 results in a file called
+``simple.pypy3-72-x86_64-linux-gnu.so``:
+
+  - ``pypy3-72`` is the ABI tag, in this case "PyPy3.x", version "7.2.x"
+
+The HPy C API is exposed to the user by including ``hpy.h`` and it is
+explained in its own section of the documentation.
+
+
+.. _hpy-target-abis:
+
+Target ABIs
+----------------
+
+Depending on the compilation options, and HPy extension can target three
+different ABIs:
+
+CPython
+  In this mode, HPy is implemented as a set of C macros and ``static inline``
+  functions which translate the HPy API into the CPython API at compile
+  time. The result is a compiled extension which is indistinguishable from a
+  "normal" one and can be distributed using all the standard tools and will
+  run at the very same speed. The ABI tag is defined by the version of CPython
+  which is used to compile it (e.g., ``cpython-37m``),
+
+HPy Universal
+  As the name suggests, the HPy Universal ABI is designed to be loaded and
```

Comment:
Worth mentioning that such packages can be distributed as wheels using the [PEP 384] stable ABI tag `abi3` (see also [the PEP 384 section of PEP 3149](https://www.python.org/dev/peps/pep-3149/#pep-384)? In any case, does this part need a PEP or would this be an import hook that adds that so signature? 

Also: the HPy Hybrid mode would force the "regular" ABI tag `cpython-37m` for the packaged wheels.

Maybe worth adding a short "from the user's perspective" section that would explain this: the CPython mode would not change anything, the HPyUniversal and HPy Hybrid would require an additional install on CPython.

Reply:
> Worth mentioning that such packages can be distributed as wheels using the [PEP 384] stable ABI tag abi3 (see also the PEP 384 section of PEP 3149? In any case, does this part need a PEP or would this be an import hook that adds that so signature?

Actually, at the moment `hpy.universal` on CPython does NOT use the limited API, e.g. we use `PyTuple_GET_ITEM` in `ctx_CallRealFunctionFromTrampoline`. We would need to run benchmarks and so how much it costs to use the limited ABI.

> Maybe worth adding a short "from the user's perspective" section that would explain this: the CPython mode would not change anything, the HPyUniversal and HPy Hybrid would require an additional install on CPython.

good idea

Reply:
@mattip I'm going to merge this PR. We can always improve this paragraph later, but at least if we merge this we put online some docs